lib/Backends/CPU/CPUBackend.cpp

```diff
@@ -133,29 +133,22 @@ CPUBackend::compileIRWithoutConstants(IRFunction *IR) const {
   return llvm::make_unique<CPUFunction>(std::move(JIT), runtimeInfo);
 }
 
-std::unique_ptr<CompiledFunction> CPUBackend::compile(Function *F) const {
-  TraceInfo traceInfo = buildManualTraceInfo(F);
-  auto IR = generateAndOptimizeIR(F, *this, shouldShareBuffers());
-  auto compiledFunc = compileIR(std::move(IR));
-  compiledFunc->setTraceInfo(std::move(traceInfo));
-  return compiledFunc;
-}
-
 std::unique_ptr<CompiledFunction>
-CPUBackend::instrumentAndCompile(Function *F) const {
+CPUBackend::compile(Function *F, CompileOptions &opts) const {
   TraceInfo traceInfo = buildManualTraceInfo(F);
   auto IR = generateAndOptimizeIR(F, *this, shouldShareBuffers());
-  autoInstrument(traceInfo, IR.get());
-  auto compiledFunc = compileIR(std::move(IR));
-  compiledFunc->setTraceInfo(std::move(traceInfo));
-  return compiledFunc;
-}
 
-std::unique_ptr<CompiledFunction>
-CPUBackend::compileWithoutConstants(Function *F) const {
-  TraceInfo traceInfo = buildManualTraceInfo(F);
-  auto IR = generateAndOptimizeIR(F, *this, shouldShareBuffers());
-  auto compiledFunc = compileIRWithoutConstants(IR.get());
+  if (opts.autoInstrument) {
+    autoInstrument(traceInfo, IR.get());
+  }
+
+  std::unique_ptr<CompiledFunction> compiledFunc;
+  if (opts.collectConstants) {
+    compiledFunc = compileIR(std::move(IR));
+  } else {
+    compiledFunc = compileIRWithoutConstants(IR.get());
```

Comment:
Do we really need different functions for `compileIR` and `compileIRWithoutConstants` or can we fold them together now and conditionally perform constants collection if needed?

Reply:
So I think ultimately we don't need both - and I did try to remove it, but due to the type signatures it's a bit complicated: Currently takes a unique_ptr to the IR (and e.g. in the Interpreter stores this unique_ptr), but collecting constants requires the Module or the IR, so we can't move it in then collect constants afterwards since we wont have either.

Reply:
Never mind then! Let's skip this for now.